usr/lib/common/object.c

```diff
@@ -795,6 +836,14 @@ CK_RV object_create_skel(STDLL_TokData_t * tokdata,
     if (rc != CKR_OK)
         goto done;
 
+    if ((class == CKO_SECRET_KEY || class == CKO_PRIVATE_KEY) &&
```

Comment:
Why not also for public keys (CKO_PUBLIC_KEY) ?
The PKEY attribute is valid for All Key objects, so then this routine should be called for ALL key objects.

Reply:
Because public keys are never used via protected keys. When e.g. performing a "verify", I take the public key attribute(s) of a given secure key and do the operation via "non-protected-key" CPACF functions. I mean, yes, we can allow CKA_IBM_PROTKEY_EXTRACTABLE = true for public keys, but they will never be used as such. Also, public keys will never get a CKA_IBM_OPAQUE_PKEY attribute.

Reply:
OK, but then your code in key_object_validate_attribute and key_object_set_default_attributes is inconsistent. key_object_set_default_attributes sets it for ALL key objects, so either that attribute is valid for ALL key objects, or not. If it is just valid for certain key objects, then use the according xxx_key_object_validate_attribute / xxx_key_object_set_default_attributes functions for checing the attribute.

And, again, I would prefer to make this token specific callback function as generic as possible, i.e. call it for ALL objects, as discussed in other comments, too.

Reply:
Done.